File: homework/ex12.6.pl
#!/usr/bin/env perl

# Exercise 12.6
#
# Using Bioperl, perform a BLAST search at the NCBI web site, then
# parse the BLAST output.
#
# I couldn't get Bioperl installed and working in a sane way, so I did
# it in Biopython instead. See: ex12.6.py

use strict;
use warnings;

use lib qw(/usr/local/share/perl5);
use lib qw(~/perl5/lib/perl5);

use Bio::Tools::Run::StandAloneBlast;
use Bio::SearchIO;
use Bio::SeqIO;

sub print_hashref {
  my ($href) = @_;
  print "HREF: $href\n";
  foreach my $key (keys %$href) {
    my $val = $$href{$key};
    print "$key: $val\n";
  }
}


# The main part.
sub main {
    my ($filename) = $ARGV;
    my $factory = Bio::Tools::Run::StandAloneBlast->new(
        -database=>'ncbi',
        -remote=>1
        );
    my $seq1 = Bio::SeqIO->new(-file=>$filename, -format=>'Fasta');
    my $report = $factory->blastn($seq1->next_seq());
    my $primary_seq = $$report{primary_seq}
}

main();

File: homework/ex12.6.py
#!/usr/bin/env python

"""Exercise 12.6

Using Biopython, perform a BLAST search at the NCBI web site, then
parse the BLAST output.

This exercise was originally intended to be implemented using
Bioperl. However, installing Bioperl was such a frustrating experience
that I chose to do this exercise in Biopython instead.

"""

import sys

import Bio.Blast.NCBIWWW
import Bio.Blast.NCBIXML
import Bio.SeqIO


def main(argv=None):
    """The main part.

    """
    (dna_filename,) = argv[1:]
    # dna_data = open(dna_file, 'rb').read()
    dna_rec = Bio.SeqIO.read(dna_filename, format='fasta')
    dna = dna_rec.seq

    print('Querying NCBI BLAST server...')
    result_xml = Bio.Blast.NCBIWWW.qblast('blastn', 'nt', dna)
    print('Parsing BLAST XML record...')
    blast_recs = Bio.Blast.NCBIXML.parse(result_xml)
    print('\n')

    for rec in blast_recs:
        print('Query:', rec.query_id)
        print('\n')
        for align in rec.alignments:
            print(align.title)
            for hsps in align.hsps:
                print(hsps)
            print('\n')


if __name__ == '__main__':
    # Do the thing
    main(argv=sys.argv)
